### ISSUE TITLE
Add ipv6 support

### ISSUE THREAD

**Issue by foguio** [haxpor/Potatso#7]
Hey @haxpor,

would it be possible to add ipv6 support?

Best,

Chris

**Comment by haxpor** [comment]
Hey thank you so much Chris for chiming in!

I believe you already tried to connect with server in ipv6 and it didn't work? I didn't know 100% in/out of Potatso's features as I came late to the game. Just want to make sure first.

**Comment by foguio** [comment]
yes i tried it but tun2socks must be updated. idk if you are able to do that :) it includes lwip which needs to be updated to version 2 at least.

also there is a problem within potatso with DNS settings. if your wifi connection drops then potatso still uses the system DNS settings which it got from wifi even when youprovide custom dns like the one from google.

**Comment by haxpor** [comment]
Thank you for lighting me up. I marked this down with feature request first.

For DNS stuff, feel free to submit a separate issue.

**Comment by hybtoy** [comment]
Just tried it. IPv6 working on Shadowsocks go version.

**Comment by foguio** [comment]
@hybtoy potatso forwards ipv6 requests to ss-go? or what do you mean? since everything within potatso is based on ipv4 that actually cant wirk?!

**Comment by hybtoy** [comment]
@foguio I have a server with ipv6 and go version of shadowsocks is installed on it. 
I just checked if ipv6 is working on latest Potatso (Swift 3) that has been built from this repository.

<center><a href="https://lh3.googleusercontent.com/-zGMQ8-J4YaM/WJw566lPUlI/AAAAAAAAANg/W7R92IvmAJE/s0/photo_2017-02-09_14-43-52.jpg"><img src="https://lh3.googleusercontent.com/-jMPOi2jpW-8/WJw57QoqyaI/AAAAAAAAANk/JyZdl-rhUhY/s0/thumb.png" border=0></a>&nbsp;&nbsp;<a href="https://lh3.googleusercontent.com/-7Hu3lmEBE1U/WJw58DLgkKI/AAAAAAAAANo/H77IU530HQk/s0/photo_2017-02-09_14-43-47.jpg"><img src="https://lh3.googleusercontent.com/-v9lvgs-wxsk/WJw58i07hwI/AAAAAAAAANs/Lbbj3YTBGY0/s0/thumb2.png" border=0></a>&nbsp;&nbsp;</center>

**Comment by foguio** [comment]
@hybtoy thats also the way your ip is checked. you need to follow the logs of SS to see if there is any ipv6 request. but they wont be because tun2socks uses or is started with ipv4.

**Comment by happyfire** [comment]
Currently I do a little test on ipv6, when running potatso from a ipv6-only network (with DNS64/NAT64 support), it fails to connect to a ipv4 server. I do a little debug, and found the NE crashed when ss-local try to bind to 127.0.0.1, the bind() return -1 because in ipv6-only network, I think the ipv4 address is invalid. Of course there are many things to do, just look at the code, case "6" is totally empty.
I found some documents to see, just mark:
[Apple's document about support ipv6](https://developer.apple.com/library/content/documentation/NetworkingInternetWeb/Conceptual/NetworkingOverview/UnderstandingandPreparingfortheIPv6Transition/UnderstandingandPreparingfortheIPv6Transition.html#//apple_ref/doc/uid/TP40010220-CH213-SW1)
[RFC4038: Application Aspects of IPv6 Transition.
](https://tools.ietf.org/html/rfc4038)

**Comment by jiangyancong** [comment]
Right now T-Mobile force all the app use ipv6 only network. Let's see what we can do to support this featue.

**Comment by aishuidedabai** [comment]
确实不支持ipv6，正在寻找解决办法，希望会的人指点一下，另外udp的处理完全是错误的，我正在尝试用pdnsd来分流
